flang/lib/Lower/IO.cpp

```diff
@@ -1401,16 +1410,94 @@ mlir::Value Fortran::lower::genReadStatement(
 
 mlir::Value Fortran::lower::genInquireStatement(
     Fortran::lower::AbstractConverter &converter,
-    const Fortran::parser::InquireStmt &) {
+    const Fortran::parser::InquireStmt &stmt) {
   auto &builder = converter.getFirOpBuilder();
   auto loc = converter.getCurrentLocation();
   mlir::FuncOp beginFunc;
-  // if (...
-  beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireUnit)>(loc, builder);
-  // else if (...
-  beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireFile)>(loc, builder);
-  // else
-  beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireIoLength)>(loc, builder);
-  // TODO: implement this!
-  llvm::report_fatal_error("INQUIRE statement is not implemented");
+  mlir::Value cookie;
+  ConditionSpecifierInfo csi{};
+
+  if (const auto *list{std::get_if<std::list<parser::InquireSpec>>(&stmt.u)}) {
+    if (hasX<Fortran::parser::FileUnitNumber>(*list)) {
+      // Determine which BeginXyz call to make.
+      beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireUnit)>(loc, builder);
+      mlir::FunctionType beginFuncTy = beginFunc.getType();
+
+      auto unit = converter.genExprValue(
+          getExprInList<Fortran::parser::FileUnitNumber>(*list), loc);
+      auto un = builder.createConvert(loc, beginFuncTy.getInput(0), unit);
+      auto file = getDefaultFilename(builder, loc, beginFuncTy.getInput(1));
+      auto line = getDefaultLineNo(builder, loc, beginFuncTy.getInput(2));
+      // Append BeginXyz call arguments.  File name and line number are always
+      // last. For a 'BeginInquireUnit' the IO runtime API looks like below
+      // BeginXyz (ExternalUnit, const char *sourceFile = nullptr, int
```

Comment:
Comment switches between BeginXyz and BeginInquireUnit.

I'm also not sure why there is so much specificity here in this comment. It's very unlikely that anyone will update this comment if the interface ever changes, for example. Also, the C++-style default arguments are just confusing, in my opinion. Since we're not lowering to a C++ compiler, they have no relevance to the FIR/MLIR/LLVM signature.

Reply:
+agreed that interfaces could changes.
My blockers were resolved when I came to know the requirements for the argument types and the number of arguments when looking at `io-api.h`.

Reply:
+agreed that interfaces could change.
My blockers were resolved when I came to know about `io-api.h` and the expected number and types of arguments.
I am ok to remove them, but at least there should be some way a new contributor must know that we need to check the runtime api files.

Reply:
I see. Maybe just simplify this comment to be a reference to the interfaces in io-api.h.